File: cses/05_range_queries/1144_salary-queries/solution.py
import sys
from bisect import bisect_left, bisect_right


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    q = int(data[idx]); idx += 1

    p = [0] * n
    for i in range(n):
        p[i] = int(data[idx]); idx += 1

    # Đọc offline toàn bộ truy vấn, gom mọi lương có thể xuất hiện để nén tọa độ:
    # gồm lương ban đầu p_i và mọi lương mới x của các truy vấn cập nhật "!".
    qtype = [0] * q     # 1 = update "!", 0 = query "?"
    qa = [0] * q
    qb = [0] * q
    all_vals = list(p)
    for j in range(q):
        t = data[idx]; idx += 1
        aa = int(data[idx]); idx += 1
        bb = int(data[idx]); idx += 1
        qa[j] = aa
        qb[j] = bb
        if t == b'!':
            qtype[j] = 1        # aa = k, bb = x (lương mới)
            all_vals.append(bb)
        else:
            qtype[j] = 0        # aa = a, bb = b (khoảng truy vấn)

    # Coordinate compression: tập giá trị duy nhất đã sắp xếp.
    vals = sorted(set(all_vals))
    m = len(vals)

    # Fenwick tree (BIT) đếm tần suất lương theo vị trí nén.
    tree = [0] * (m + 1)

    # cur[k] = vị trí nén (1-indexed) của lương hiện tại của nhân viên k.
    cur = [0] * (n + 1)

    # Khởi tạo BIT: cộng 1 vào vị trí nén của lương ban đầu từng nhân viên.
    for i in range(n):
        pos = bisect_left(vals, p[i]) + 1
        cur[i + 1] = pos
        # update(pos, +1): lan lên các nút cha theo j & -j.
        j = pos
        while j <= m:
            tree[j] += 1
            j += j & (-j)

    out = []
    ap = out.append
    for jq in range(q):
        a = qa[jq]
        b = qb[jq]
        if qtype[jq] == 1:
            # "! k x": rút 1 khỏi vị trí lương cũ rồi cộng 1 vào vị trí lương mới.
            k = a
            x = b
            # update(cur[k], -1)
            j = cur[k]
            while j <= m:
                tree[j] -= 1
                j += j & (-j)
            pos = bisect_left(vals, x) + 1
            cur[k] = pos
            # update(pos, +1)
            j = pos
            while j <= m:
                tree[j] += 1
                j += j & (-j)
        else:
            # "? a b": đáp án = prefix(hi) - prefix(lo).
            hi = bisect_right(vals, b)   # số lương <= b
            lo = bisect_left(vals, a)    # số lương < a
            s = 0
            j = hi
            while j > 0:
                s += tree[j]
                j -= j & (-j)
            j = lo
            while j > 0:
                s -= tree[j]
                j -= j & (-j)
            ap(s)

    sys.stdout.buffer.write(('\n'.join(map(str, out)) + ('\n' if out else '')).encode())


main()

File: cses/05_range_queries/1144_salary-queries/solution.cpp
#include <algorithm>
#include <cstddef>
#include <iostream>
#include <vector>

// Lưu lại từng truy vấn để xử lí offline: type là '!' (cập nhật) hoặc '?' (đếm).
struct Query {
    char type;
    int a;
    int b;
};

// Fenwick tree (BIT) duy trì tần suất lương theo vị trí nén, hỗ trợ
// point-update và prefix-sum, mỗi thao tác O(log N).
class FenwickTree {
public:
    explicit FenwickTree(std::size_t size) : tree_(size + 1, 0) {}

    // Cộng delta vào một vị trí rồi lan lên các nút cha theo i & -i.
    void add(std::size_t position, int delta) {
        const std::size_t size = tree_.size() - 1;
        for (std::size_t i = position; i <= size; i += i & -i) {
            tree_[i] += delta;
        }
    }

    // Tổng tiền tố [1, position].
    [[nodiscard]] int prefix_sum(std::size_t position) const {
        int result = 0;
        for (std::size_t i = position; i > 0; i -= i & -i) {
            result += tree_[i];
        }
        return result;
    }

private:
    std::vector<int> tree_;
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::size_t n = 0;
    std::size_t q = 0;
    std::cin >> n >> q;

    // Đọc lương ban đầu, đồng thời gom vào tập giá trị cần nén.
    std::vector<int> salary(n);
    std::vector<int> values;
    values.reserve(n + q);
    for (int& value : salary) {
        std::cin >> value;
        values.push_back(value);
    }

    // Đọc offline mọi truy vấn; với "! k x" gom thêm lương mới x vào tập nén.
    std::vector<Query> queries;
    queries.reserve(q);
    for (std::size_t i = 0; i < q; ++i) {
        Query query{};
        std::cin >> query.type >> query.a >> query.b;
        queries.push_back(query);
        if (query.type == '!') {
            values.push_back(query.b);
        }
    }

    // Coordinate compression: sắp xếp và lấy tập giá trị duy nhất.
    std::sort(values.begin(), values.end());
    values.erase(std::unique(values.begin(), values.end()), values.end());

    // Đổi một mức lương sang vị trí nén 1-indexed (BIT đánh chỉ số từ 1).
    const auto compressed_position = [&values](int value) {
        return static_cast<std::size_t>(
                   std::lower_bound(values.begin(), values.end(), value) - values.begin()) +
               1;
    };

    // Khởi tạo BIT: cộng 1 vào vị trí nén của lương ban đầu từng nhân viên,
    // lưu current_position[i] là vị trí nén hiện tại của nhân viên i.
    FenwickTree frequencies(values.size());
    std::vector<std::size_t> current_position(n);
    for (std::size_t i = 0; i < n; ++i) {
        current_position[i] = compressed_position(salary[i]);
        frequencies.add(current_position[i], 1);
    }

    for (const Query& query : queries) {
        if (query.type == '!') {
            // Cập nhật lương: rút 1 khỏi vị trí cũ, đặt vị trí mới rồi cộng 1.
            const std::size_t employee = static_cast<std::size_t>(query.a - 1);
            frequencies.add(current_position[employee], -1);
            current_position[employee] = compressed_position(query.b);
            frequencies.add(current_position[employee], 1);
        } else {
            // Đếm số lương trong [a, b] = prefix(số lương <= b) - prefix(số lương < a);
            // dùng bisect quy a, b về biên chỉ số hợp lệ trên tập đã nén.
            const std::size_t below_lower = static_cast<std::size_t>(
                std::lower_bound(values.begin(), values.end(), query.a) - values.begin());
            const std::size_t at_most_upper = static_cast<std::size_t>(
                std::upper_bound(values.begin(), values.end(), query.b) - values.begin());
            std::cout << frequencies.prefix_sum(at_most_upper) -
                             frequencies.prefix_sum(below_lower)
                      << '\n';
        }
    }
    return 0;
}
